### ISSUE TITLE
DNX120 Interlaced not decoded correctly

### ISSUE THREAD

**Issue by FranceBB** [mpv-player/mpv#9157]
Hi there,
DNX120 indexing is broken in:

- LWLibavVideoSource
- FFVideoSource
- DirectShowSource (LAV)
- FFMpeg
- Davinci Resolve

The following players also fail to decode it properly:

- MPV
- MPC-HC (LAV)
- PotPlayer

The following software decode it correctly:

- AVID Media Composer
- AVID Media Director
- Adobe Premiere
- EVS Hardware Playout Ports

Here's the mediainfo:

![JFQWYwr](https://user-images.githubusercontent.com/18946343/130841922-bb0df5a5-2192-47fe-b8d9-aba0f9f6e9df.png)

[https://i.imgur.com/JFQWYwr.png](https://i.imgur.com/JFQWYwr.png)

This is how it's decoded by MPV by default:

![UdEp4Nt](https://user-images.githubusercontent.com/18946343/130842108-0140d1d7-f092-4a44-86fd-aaf24339052d.png)

[https://i.imgur.com/UdEp4Nt.png](https://i.imgur.com/UdEp4Nt.png)

And this is how it should be instead:

![dfPB74v](https://user-images.githubusercontent.com/18946343/130842163-2aa13190-40d2-4fd4-8edc-e326c26dba35.png)

[https://i.imgur.com/dfPB74v.png](https://i.imgur.com/dfPB74v.png)

Hardware acceleration/decoding is disabled, everything goes through software decoding.

How to reproduce: just open the file.

Same happens in ffplay:

```
[FranceBB@router-localhost ~]$ ffplay -i "/home/FranceBB/Downloads/20210429WiganvHullFCX_C921h25m29s09.mxf"
ffplay version 4.4 Copyright (c) 2003-2021 the FFmpeg developers
built with gcc 11 (GCC)
configuration: --prefix=/usr --bindir=/usr/bin --datadir=/usr/share/ffmpeg --docdir=/usr/share/doc/ffmpeg --incdir=/usr/include/ffmpeg --libdir=/usr/lib64 --mandir=/usr/share/man --arch=x86_64 --optflags='-O2 
-flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 
-fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection' --extra-ldflags='-Wl,-z,relro -Wl,--as-needed 
-Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld ' --extra-cflags=' -I/usr/include/rav1e' --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libvo-amrwbenc --enable-version3 
--enable-bzlib --disable-crystalhd --enable-fontconfig --enable-frei0r --enable-gcrypt --enable-gnutls --enable-ladspa --enable-libaom --enable-libdav1d --enable-libass --enable-libbluray --enable-libcdio 
--enable-libdrm --enable-libjack --enable-libfreetype --enable-libfribidi --enable-libgsm --enable-libmp3lame --enable-libmysofa --enable-nvenc --enable-openal --enable-opencl --enable-opengl 
--enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librsvg --enable-librav1e --enable-libsmbclient --enable-version3 --enable-libsoxr --enable-libspeex --enable-libsrt 
--enable-libssh --enable-libsvtav1 --enable-libtheora --enable-libvorbis --enable-libv4l2 --enable-libvidstab --enable-libvmaf --enable-version3 --enable-vapoursynth --enable-libvpx --enable-vulkan 
--enable-libglslang --enable-libx264 --enable-libx265 --enable-libxvid --enable-libxml2 --enable-libzimg --enable-libzvbi --enable-lv2 --enable-avfilter --enable-avresample --enable-libmodplug 
--enable-postproc --enable-pthreads --disable-static --enable-shared --enable-gpl --disable-debug --disable-stripping --shlibdir=/usr/lib64 --enable-lto --enable-libmfx --enable-runtime-cpudetect
libavutil 56. 70.100 / 56. 70.100
libavcodec 58.134.100 / 58.134.100
libavformat 58. 76.100 / 58. 76.100
libavdevice 58. 13.100 / 58. 13.100
libavfilter 7.110.100 / 7.110.100
libavresample 4. 0. 0 / 4. 0. 0
libswscale 5. 9.100 / 5. 9.100
libswresample 3. 9.100 / 3. 9.100
libpostproc 55. 9.100 / 55. 9.100
Input #0, mxf, from '/home/FranceBB/Downloads/20210429WiganvHullFCX_C921h25m29s09.mxf':
Metadata:
operational_pattern_ul: 060e2b34.04010101.0d010201.01010100
application_platform: Microsoft Windows 7 Professional Service Pack 1 (Build 7601)
uid : f3687fb0-a928-11eb-92b0-0cc47a8346df
generation_uid : f3687fb0-a928-11eb-92b1-0cc47a8346df
company_name : OpenCube
product_name : MXFTk Advanced
product_uid : 3a4fe380-0d01-11e4-869f-3cd92b5c1dfc
product_version : 2.10.4.20180828
product_version_num: 2.10.4.0.1
toolkit_version_num: 2.10.4.0.1
modification_date: 2021-04-29T21:25:14.052000Z
material_package_umid: 0x060A2B34010101050101052013000000F36858A0A92811EB92A40CC47A8346DF
timecode : 21:25:24:09
Duration: 00:00:35.68, start: 0.000000, bitrate: 121346 kb/s
Stream #0:0: Video: dnxhd (DNXHD), yuv422p(bt709/unknown/bt709, top first), 1920x1080, SAR 1:1 DAR 16:9, 25 fps, 25 tbr, 25 tbn, 25 tbc
Metadata:
file_package_umid: 0x060A2B34010101050101052013000000F3683190A92811EB929C0CC47A8346DF
track_name : Picture
5.24 M-V: 0.045 fd= 0 aq= 0KB vq=15392KB sq= 0B f=0/0
```

-thread_type slice actually solves the issue in FFMpeg and FFPlay:

`ffplay -i "/home/FranceBB/Downloads/20210429WiganvHullFCX_C921h25m29s09.mxf" -thread_type slice`

There are also some DNX120 that play ok from some cameras and some other DNX120 that have green fields from some totally different cameras.
Going deeper in the analysis, those from the camera brands that are decoded correctly have:

`[dnxhd @ 000001f4fd9d7c80] interlaced 3, cur field 1`

those from the camera brands that are decoded incorrectly (unless -thread_type slice is specified) have:

`[dnxhd @ 0000013fc0277c80] interlaced 2, cur field 0`

the values are bits in the frame header, what i don't get is why we have the entries 2 times

```
[dnxhd @ 0000025bc07c7c80] interlaced 2, cur field 0
[dnxhd @ 0000025bc07c7c80] Profile cid 1242.
[dnxhd @ 0000025bc07c7c80] 1920x1080, 4:2:2 8 bits, MBAFF=0 ACT=0
[dnxhd @ 0000025bc07c7c80] interlaced 2, cur field 0
[dnxhd @ 0000025bc07c7c80] 1920x1080, 4:2:2 8 bits, MBAFF=0 ACT=0
```

in the cameras that produce the version with the decoding issue, it's 2 times "interlaced 2, cur field 0" while in the camera brands that produce a file correctly decoded without any flag, only the first entry is the same while the second entry is

`[dnxhd @ 00000222415e7c80] interlaced 3, cur field 1`

I've been testing several DNX120 and those that are decoded correctly in FFMpeg without the -thread_type slice have:

```
[dnxhd @ 0000025bc07c7c80] interlaced 2, cur field 0
[dnxhd @ 00000222415e7c80] interlaced 3, cur field 1
```

while the ones that are decoded incorrectly and need the -thread_type slice in order to decode them correctly have:

```
[dnxhd @ 0000025bc07c7c80] interlaced 2, cur field 0
[dnxhd @ 0000025bc07c7c80] interlaced 2, cur field 0
```

Besides, if I use ffprobe to see what's going on in the files that are not decoded correctly I get:

```
ffprobe "test.dnxhd" -show_frames |findstr top_field_first
top_field_first=0
top_field_first=1
top_field_first=0
top_field_first=1
top_field_first=0
top_field_first=1
top_field_first=0
top_field_first=1
top_field_first=0
top_field_first=1
top_field_first=0
top_field_first=1
```

Now, according to the DNX standard, we have one frame header per field (unlike in other codecs like MPEG-2 where we only have one frame header for each pair of fields).
The frame header start sequence is 00 00 02 80 01 - the next byte tells us if we deal with the first or second field of an interlaced frame
00 00 02 80 01 02 indicates first field
00 00 02 80 01 03 indicates second field
If I take one of the DNX120 which is decoded correctly in ffmpeg by default, I have 02 and 03 that change with every frame header, like so:

```
frame 0, first field 00 00 02 80 01 02
frame 0, second field 00 00 02 80 01 03
00 00 02 80 01 02
00 00 02 80 01 03
00 00 02 80 01 02
00 00 02 80 01 03
```

but this is not the case for the DNX120 files we have received from the pitch in which we have:

```
00 00 02 80 01 02
00 00 02 80 01 02
00 00 02 80 01 03
00 00 02 80 01 03
00 00 02 80 01 02
00 00 02 80 01 02
00 00 02 80 01 03
00 00 02 80 01 03
```

using gsar https://trac.ffmpeg.org/attachment/ticket/9255/gsar.exe to modify the header works and makes the files possible to decode/index like so:

`type "INPUTFILE" | C:\temp\gsar.exe -s:x83:x09:x40:x00:x00:x00:x02:x80:x01:x03 -r:x83:x09:x40:x00:x00:x00:x02:x80:x01:x02 -F | C:\temp\gsar.exe -s:x60:x0D:xC0:xDE:x00:x00:x02:x80:x01:x02 -r:x60:x0D:xC0:xDE:x00:x00:x02:x80:x01:x03 -F > "OUTPUTFILE"`

this also fixes the decoding issue and it can be executed on Windows on the original file before decoding it:

```
cat.exe "INPUTFILE" | gsar.exe -s:x83:x09:x40:x00:x00:x00:x02:x80:x01:x03 -r:x83:x09:x40:x00:x00:x00:x02:x80:x01:x02 -F  | 
gsar.exe -s:x60:x0D:xC0:xDE:x00:x00:x02:x80:x01:x02 -r:x60:x0D:xC0:xDE:x00:x00:x02:x80:x01:x03 -F > "OUTPUTFILE"
```

gsar.exe https://trac.ffmpeg.org/attachment/ticket/9255/gsar.exe
cat.exe https://trac.ffmpeg.org/attachment/ticket/9255/cat.exe

So, in a nutshell, I believe that MPV should set -thread_type slice whenever there's a DNX120 interlaced to provide the correct decoding.
The developers of LWLibavVideoSource (LSMASH) as Avisynth indexer did something similar to address the issue.
They did so 'cause I raised the issue over there as well along with the ffmpeg bug tracker which you can find here: https://forum.doom9.org/showthread.php?t=182885 and https://trac.ffmpeg.org/ticket/9255#ticket

Sample files (valid for 7 days): https://we.tl/t-J4oRp0JFUA

**Comment by FranceBB** [comment]
Yes it works in MPV now that i has been fixed in FFMpeg. ;
![Screenshot from 2021-09-24 15-53-40](https://user-images.githubusercontent.com/18946343/134696255-d79af5bb-6ca6-451e-8e86-680fca245ae1.png)
)
